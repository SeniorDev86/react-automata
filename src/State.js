import React from 'react'
import PropTypes from 'prop-types'

class State extends React.Component {
  render() {
    return this.context.machineState === this.props.name
      ? this.props.children
      : null
  }
}

State.defaultProps = {
  children: null,
}

State.propTypes = {
  name: PropTypes.string.isRequired,
  children: PropTypes.node,
}

State.contextTypes = {
  machineState: PropTypes.string,
}

export default State
