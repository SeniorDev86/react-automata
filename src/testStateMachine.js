import React from 'react'
import PropTypes from 'prop-types'
import TestRenderer from 'react-test-renderer'

const createContext = (context, Component) => {
  class Context extends React.Component {
    getChildContext() {
      return { ...context }
    }

    render() {
      return <Component />
    }
  }

  Context.childContextTypes = {
    machineState: PropTypes.string,
  }

  return Context
}

const injectState = (renderer, Component, fixtures) => {
  if (fixtures) {
    const { instance } = renderer.root.findByType(Component)
    instance.setState(fixtures)
  }
}

const moveToNextState = (config, Component, machineState) => {
  const { on: actions } = config.machine.states[machineState]
  if (actions) {
    Object.values(actions).forEach(state => {
      toMatchSnapshot(config, Component, state)
    })
  }
}

const toMatchSnapshot = (config, Component, machineState) => {
  const Context = createContext({ machineState }, Component)
  const renderer = TestRenderer.create(<Context />)
  injectState(renderer, Component, config.fixtures[machineState])
  expect(renderer.toJSON()).toMatchSnapshot(machineState)
  moveToNextState(config, Component, machineState)
}

const testStateMachine = (config, Component) => {
  toMatchSnapshot(config, Component, config.machine.initial)
}

export default testStateMachine
