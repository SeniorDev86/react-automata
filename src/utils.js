export const getComponentName = Component =>
  Component.displayName || Component.name || 'Component'

export const isStateless = Component =>
  !(Component.prototype && Component.prototype.isReactComponent)
