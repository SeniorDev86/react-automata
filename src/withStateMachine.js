import React from 'react'
import PropTypes from 'prop-types'
import { Machine } from 'xstate'

const withStateMachine = config => Component => {
  class StateMachine extends React.Component {
    state = {
      machineState: this.machine.getInitialState(),
    }

    getChildContext() {
      return { ...this.state }
    }

    machine = Machine(config)

    handleTransition = action => {
      this.setState(prevState => ({
        machineState: this.machine.transition(prevState.machineState, action)
          .value,
      }))
    }

    render() {
      return <Component {...this.props} transition={this.handleTransition} />
    }
  }

  StateMachine.childContextTypes = {
    machineState: PropTypes.string,
  }

  return StateMachine
}

export default withStateMachine
