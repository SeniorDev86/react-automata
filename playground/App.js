import React from 'react'
import { State, withStateMachine } from '../src'

export const machine = {
  initial: 'idle',
  states: {
    idle: {
      on: {
        FETCH: 'fetching',
      },
    },
    fetching: {
      on: {
        SUCCESS: 'success',
        ERROR: 'error',
      },
    },
    success: {},
    error: {
      on: {
        FETCH: 'fetching',
      },
    },
  },
}

export class App extends React.Component {
  componentWillTransition(event) {
    if (event === 'FETCH') {
      fetch('https://api.github.com/users/gaearon/gists')
        .then(response => response.json())
        .then(gists => this.props.transition('SUCCESS', { gists }))
        .catch(() => this.props.transition('ERROR'))
    }
  }

  handleClick = () => {
    this.props.transition('FETCH')
  }

  render() {
    return (
      <div>
        <h1>State Machine</h1>
        <State value={['idle', 'error']}>
          <button onClick={this.handleClick}>
            {this.props.machineState === 'idle' ? 'Fetch' : 'Retry'}
          </button>
        </State>
        <State value="fetching">Loading...</State>
        <State value="success">
          <ul>
            {this.props.gists
              .filter(gist => gist.description)
              .map(gist => <li key={gist.id}>{gist.description}</li>)}
          </ul>
        </State>
        <State value="error">Oh, snap!</State>
      </div>
    )
  }
}

const options = {
  devTools: true,
  initialData: { gists: [] },
}

export default withStateMachine(machine, options)(App)
