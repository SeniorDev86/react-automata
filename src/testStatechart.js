import React from 'react'
import TestRenderer from 'react-test-renderer'
import { Machine } from 'xstate'
import { getShortestPaths } from 'xstate/lib/graph'
import idx from 'idx'
import invariant from 'invariant'
import { getContextValue } from './utils'
import withStatechart from './withStatechart'

const testStatechart = (config, Component) => {
  invariant(
    !Component.isStateMachine,
    `It seems you are testing a component wrapped into \`withStatechart\`, please use a base component instead.
    See https://github.com/MicheleBertoli/react-automata/issues/46`
  )

  const { channel, statechart } = config
  const machine = Machine(statechart)
  const paths = getShortestPaths(machine)

  Object.keys(paths).forEach(key => {
    const initialData = idx(config, _ => _.fixtures.initialData)
    const StateMachine = withStatechart(statechart, { channel })(Component)
    const renderer = TestRenderer.create(
      <StateMachine initialData={initialData} />
    )
    const instance = renderer.getInstance()

    paths[key].forEach(({ event, state }) => {
      const fixtures = idx(config, _ => _.fixtures[state][event])

      instance.handleTransition(event, fixtures)
    })

    const { machineState } = getContextValue(
      instance.getChildContext(),
      channel
    )

    expect(renderer.toJSON()).toMatchSnapshot(machineState)
  })
}

export default testStatechart
