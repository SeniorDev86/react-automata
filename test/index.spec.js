import { testStateMachine } from '../src'
import App from './app'

const machine = {
  initial: 'idle',
  states: {
    idle: {
      on: {
        FETCH: 'fetching',
      },
    },
    fetching: {
      on: {
        SUCCESS: 'success',
        ERROR: 'error',
      },
    },
    success: {},
    error: {},
  },
}

const fixtures = {
  success: {
    gists: [
      {
        id: 'ID1',
        description: 'GIST1',
      },
      {
        id: 'ID2',
        description: 'GIST2',
      },
    ],
  },
}

test('it works', () => {
  testStateMachine({ machine, fixtures }, App)
})
