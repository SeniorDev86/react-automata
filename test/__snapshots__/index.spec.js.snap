// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error 1`] = `
<div>
  <h1>
    State Machine
  </h1>
  Oh, snap!
</div>
`;

exports[`fetching 1`] = `
<div>
  <h1>
    State Machine
  </h1>
  Loading...
</div>
`;

exports[`idle 1`] = `
<div>
  <h1>
    State Machine
  </h1>
  <button
    onClick={[Function]}
  >
    Fetch
  </button>
</div>
`;

exports[`success 1`] = `
<div>
  <h1>
    State Machine
  </h1>
  <ul>
    <li>
      GIST1
    </li>
    <li>
      GIST2
    </li>
  </ul>
</div>
`;
